File: mobile/backend/app/schemas/product.py
"""
Product schemas for request/response validation.
"""

from typing import Optional
from pydantic import BaseModel, Field


class ProductCreate(BaseModel):
    """Schema for creating a product."""
    sku: str = Field(..., min_length=1)
    nom_produit: str = Field(..., min_length=1)
    unite_mesure: str = Field(default="pcs")
    categorie: Optional[str] = None
    actif: bool = True
    colisage_fardeau: Optional[int] = Field(default=None, ge=0)
    colisage_palette: Optional[int] = Field(default=None, ge=0)
    volume_pcs: Optional[float] = Field(default=None, ge=0)
    poids: Optional[float] = Field(default=None, ge=0)
    is_gerbable: bool = False
    demand_freq: float = Field(default=0.0, ge=0)
    reception_freq: float = Field(default=0.0, ge=0)
    delivery_freq: float = Field(default=0.0, ge=0)


class ProductUpdate(BaseModel):
    """Schema for updating a product."""
    nom_produit: Optional[str] = Field(default=None, min_length=1)
    unite_mesure: Optional[str] = None
    categorie: Optional[str] = None
    actif: Optional[bool] = None
    colisage_fardeau: Optional[int] = Field(default=None, ge=0)
    colisage_palette: Optional[int] = Field(default=None, ge=0)
    volume_pcs: Optional[float] = Field(default=None, ge=0)
    poids: Optional[float] = Field(default=None, ge=0)
    is_gerbable: Optional[bool] = None
    demand_freq: Optional[float] = Field(default=None, ge=0)
    reception_freq: Optional[float] = Field(default=None, ge=0)
    delivery_freq: Optional[float] = Field(default=None, ge=0)


class ProductResponse(BaseModel):
    """Schema for product response."""
    id: str
    sku: str
    nom_produit: str
    unite_mesure: str
    categorie: Optional[str] = None
    actif: bool
    colisage_fardeau: Optional[int] = None
    colisage_palette: Optional[int] = None
    volume_pcs: Optional[float] = None
    poids: Optional[float] = None
    is_gerbable: bool
    demand_freq: float
    reception_freq: float
    delivery_freq: float
    created_at: Optional[str] = None
    updated_at: Optional[str] = None

    class Config:
        from_attributes = True

File: mobile/frontend/lib/models/product.dart
class Product {
  final String id;
  final String sku;
  final String nomProduit;
  final String uniteMesure;
  final String? categorie;
  final bool actif;
  final int? colisageFardeau;
  final int? colisagePalette;
  final double? volumePcs;
  final double? poids;
  final bool isGerbable;
  final double demandFreq;
  final double receptionFreq;
  final double deliveryFreq;
  final String? createdAt;
  final String? updatedAt;

  Product({
    required this.id,
    required this.sku,
    required this.nomProduit,
    this.uniteMesure = 'pcs',
    this.categorie,
    this.actif = true,
    this.colisageFardeau,
    this.colisagePalette,
    this.volumePcs,
    this.poids,
    this.isGerbable = false,
    this.demandFreq = 0.0,
    this.receptionFreq = 0.0,
    this.deliveryFreq = 0.0,
    this.createdAt,
    this.updatedAt,
  });

  factory Product.fromJson(Map<String, dynamic> json) {
    return Product(
      id: json['id'] ?? '',
      sku: json['sku'] ?? '',
      nomProduit: json['nom_produit'] ?? '',
      uniteMesure: json['unite_mesure'] ?? 'pcs',
      categorie: json['categorie'],
      actif: json['actif'] ?? true,
      colisageFardeau: json['colisage_fardeau'],
      colisagePalette: json['colisage_palette'],
      volumePcs: (json['volume_pcs'] as num?)?.toDouble(),
      poids: (json['poids'] as num?)?.toDouble(),
      isGerbable: json['is_gerbable'] ?? false,
      demandFreq: (json['demand_freq'] as num?)?.toDouble() ?? 0.0,
      receptionFreq: (json['reception_freq'] as num?)?.toDouble() ?? 0.0,
      deliveryFreq: (json['delivery_freq'] as num?)?.toDouble() ?? 0.0,
      createdAt: json['created_at'],
      updatedAt: json['updated_at'],
    );
  }

  Map<String, dynamic> toJson() => {
    'sku': sku,
    'nom_produit': nomProduit,
    'unite_mesure': uniteMesure,
    'categorie': categorie,
    'actif': actif,
    'colisage_fardeau': colisageFardeau,
    'colisage_palette': colisagePalette,
    'volume_pcs': volumePcs,
    'poids': poids,
    'is_gerbable': isGerbable,
    'demand_freq': demandFreq,
    'reception_freq': receptionFreq,
    'delivery_freq': deliveryFreq,
  };
}
